### DISCUSSION TITLE
Disable SQL injection syntax

### DISCUSSION THREAD

**Discussion by robertlagrant** [sqlalchemy/sqlalchemy#7296] [Ideas]
The idea is to disable (or make configurable via env var) the ability for SQL statements to be executed via `connection.execute()` when they have parameters in the statement supplied directly in a raw SQL string. 

E.g. throw a `ValueError` when a SQL string is supplied with any `WHERE` parameters are supplied.

The purpose of this is to attempt to limit the ability of developers such as myself to accidentally write (via f-strings or similar) code that is vulnerable to SQL injection.

I don't know if this appropriate to be done as as core SQLAlchemy feature, or if it would be better as some sort of plugin. If the latter I'd be keen to write it, with perhaps a little guidance on the best way to implement.

Thanks!

Rob

**Comment by CaselIT** [comment]
Hi,

Using the 2.0 syntax `execute` will error if executing a string directly, it needs to be converted to a `text` object.

Also can you provide an example of what you mean by
> when a SQL string is supplied with any WHERE parameters are supplied.

doing `select(...).where(column == 'some string')` is generally safe, since the string is passed as a parameter

**Comment by zzzeek** [comment]
we use the ``text()`` construct to execute strings.  you have two options for where you can parse this and throw a ValueError.  Most efficient would be as a [@compiles](https://docs.sqlalchemy.org/en/14/core/compiler.html) recipe, because then the work you do to parse and validate the string would be cached and not have to run on every execution (this makes it much more viable as a plugin that would be of general use). 

Another hook you could use is [before_cursor_execute](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=before_cursor_execute#sqlalchemy.events.ConnectionEvents.before_cursor_execute) which runs for every statement execution, including connection.execute() and connection.exec_driver_sql().   (Edited)

**Reply by zzzeek** [reply]
i was wrong about before_cursor_execute() not emitting fro exec_driver_sql() (thankfully), it counts for that method also.  so before_cursor_execute() will intercept all SQL statements.
